tuf/ngclient/updater.py

```diff
@@ -368,210 +372,80 @@ def _load_targets(self, role: str, parent_role: str) -> None:
             self._trusted_set.update_delegated_targets(data, role, parent_role)
             self._persist_metadata(role, data)
 
-    def _preorder_depth_first_walk(self, target_filepath) -> Dict:
+    def _preorder_depth_first_walk(
+        self, target_filepath: str
+    ) -> Union[Dict[str, Any], None]:
         """
         Interrogates the tree of target delegations in order of appearance
         (which implicitly order trustworthiness), and returns the matching
         target found in the most trusted role.
         """
 
-        target = None
-        role_names = [("targets", "root")]
-        visited_role_names = set()
+        # List of delegations to be interrogated. A (role, parent role) pair
+        # is needed to load and verify the delegated targets metadata.
+        delegations_to_visit = [("targets", "root")]
+        visited_role_names: Set[Tuple[str, str]] = set()
         number_of_delegations = self.config.max_delegations
 
         # Preorder depth-first traversal of the graph of target delegations.
-        while (
-            target is None and number_of_delegations > 0 and len(role_names) > 0
-        ):
+        while number_of_delegations > 0 and len(delegations_to_visit) > 0:
 
             # Pop the role name from the top of the stack.
-            role_name, parent_role = role_names.pop(-1)
-            self._load_targets(role_name, parent_role)
+            role_name, parent_role = delegations_to_visit.pop(-1)
+
             # Skip any visited current role to prevent cycles.
             if (role_name, parent_role) in visited_role_names:
                 logger.debug("Skipping visited current role %s", role_name)
                 continue
 
             # The metadata for 'role_name' must be downloaded/updated before
             # its targets, delegations, and child roles can be inspected.
+            self._load_targets(role_name, parent_role)
 
-            role_metadata = self._trusted_set[role_name].signed
+            role_metadata: Targets = self._trusted_set[role_name].signed
             target = role_metadata.targets.get(target_filepath)
```

Comment:
```suggestion
            target: Union[TargetFile, None] = role_metadata.targets.get(target_filepath)
```
According to the metadata API if `role_metadata` is an instance of `Targets`, then 
`role_metadata.targets.get(target_filepath)` should be a `TargetFile` if it exists.
I think it make sense to add annotation here.

Reply:
(as reply to Martins comment:) since role_metadata is now a `Targets`, this does not need an annotation for target: static type check already figures out that target is a `TargetFile | None`